RecoMET/METPUSubtraction/plugins/DeepMETProducer.cc

```diff
@@ -0,0 +1,162 @@
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+
+#include "DataFormats/PatCandidates/interface/MET.h"
+#include "DataFormats/PatCandidates/interface/PackedCandidate.h"
+
+#include "PhysicsTools/TensorFlow/interface/TensorFlow.h"
+
+struct DeepMETCache {
+  std::atomic<tensorflow::GraphDef*> graph_def;
+};
+
+class DeepMETProducer : public edm::stream::EDProducer<edm::GlobalCache<DeepMETCache> > {
+public:
+  explicit DeepMETProducer(const edm::ParameterSet&, const DeepMETCache*);
+  void produce(edm::Event& event, const edm::EventSetup& setup) override;
+  static void fillDescriptions(edm::ConfigurationDescriptions& descriptions);
+
+  // static methods for handling the global cache
+  static std::unique_ptr<DeepMETCache> initializeGlobalCache(const edm::ParameterSet&);
+  static void globalEndJob(DeepMETCache*);
+
+private:
+  const edm::EDGetTokenT<std::vector<pat::PackedCandidate> > pf_token_;
+  float norm_;
+  bool ignore_leptons_;
+  unsigned int max_n_pf_;
+
+  tensorflow::Session* session_;
+
+  std::unordered_map<int, int32_t> charge_embedding_;
+  std::unordered_map<int, int32_t> pdg_id_embedding_;
+};
+
+namespace {
+  float divide_and_rm_outlier(float val, float norm) {
+    float ret_val = val / norm;
+    if (ret_val > 1e6 || ret_val < -1e6)
+      return 0.;
+    return ret_val;
+  }
+}  // namespace
+
+DeepMETProducer::DeepMETProducer(const edm::ParameterSet& cfg, const DeepMETCache* cache)
+    : pf_token_(consumes<std::vector<pat::PackedCandidate> >(cfg.getParameter<edm::InputTag>("pf_src"))),
+      norm_(cfg.getParameter<double>("norm_factor")),
+      ignore_leptons_(cfg.getParameter<bool>("ignore_leptons")),
+      max_n_pf_(cfg.getParameter<unsigned int>("max_n_pf")) {
+  session_ = tensorflow::createSession(cache->graph_def);
+  produces<pat::METCollection>();
+  charge_embedding_ = {{-1, 0}, {0, 1}, {1, 2}};
+  pdg_id_embedding_ = {
+      {-211, 0}, {-13, 1}, {-11, 2}, {0, 3}, {1, 4}, {2, 5}, {11, 6}, {13, 7}, {22, 8}, {130, 9}, {211, 10}};
+  ;
+}
+
+void DeepMETProducer::produce(edm::Event& event, const edm::EventSetup& setup) {
+  edm::Handle<std::vector<pat::PackedCandidate> > pf_h;
+  event.getByToken(pf_token_, pf_h);
+
+  // PF keys [b'PF_dxy', b'PF_dz', b'PF_eta', b'PF_mass', b'PF_pt', b'PF_puppiWeight', b'PF_px', b'PF_py']
+  tensorflow::TensorShape shape({1, max_n_pf_, 8});
+  tensorflow::TensorShape cat_shape({1, max_n_pf_, 1});
+  tensorflow::Tensor input(tensorflow::DT_FLOAT, shape);
+  tensorflow::Tensor input_cat0(tensorflow::DT_FLOAT, cat_shape);
+  tensorflow::Tensor input_cat1(tensorflow::DT_FLOAT, cat_shape);
+  tensorflow::Tensor input_cat2(tensorflow::DT_FLOAT, cat_shape);
+
+  tensorflow::TensorShape out_shape({1, 2});
+  tensorflow::Tensor output(tensorflow::DT_FLOAT, out_shape);
+
+  tensorflow::NamedTensorList input_list = {
+      {"input", input}, {"input_cat0", input_cat0}, {"input_cat1", input_cat1}, {"input_cat2", input_cat2}};
```

Comment:
static const?
Please check other values if can be made const, if not constexpr.

Reply:
Couldn't be static const because the tensors were defined and constructed in the local scope. I've now moved the tensors to the class level, with the advantage that this avoids reconstructing them in each event but will keep memory occupied (so I'm actually not sure it if's better or not).

Reply:
ah, sorry, I was looking apparently at it and seeing only the format definition and missed the obvious second argument in each case, which are per-event.
Local defs should be fine as well as data members.